### DISCUSSION TITLE
Improve experience of `spin init`

### DISCUSSION THREAD

**Discussion by jaydrogers** [serversideup/spin#96] [Ideas]
# 👉 Describe the problem
[comment]: <> (What are you trying to solve?)
- Spin Pro does not support `spin init` yet
- Some say they are trying to init in an existing project, but they run into PHP version issues or missing extensions

# 🏆 How to solve this problem
[comment]: <> (Describe the feature that you are proposing and how it will solve the problem)

### Allow users to run `spin init`
Users should be able to run:

```
spin init laravel-pro
```
This should walk them through the process of getting Spin configured.

### ⚠️ Challenges
Initializing in an existing project comes with challenges:
- Users might be on an old version of PHP
- They might have specific extensions that need to be installed that we cannot automatically
- Users may need to perform a lot of .env updates to work with the new spin infrastructure
- Users might have existing Dockerfiles that need to be removed/updated

### Update the on-boarding screen
1. Choose PHP version
2. Ask if they have custom extensions
3. Detect if they have existing Dockerfiles
4. Give them documentation if they need to take any manual steps

# 🗣️ Further discussion
- Do you think the whole process should be automatic? Or do you think it should be manual?
- What other issues could we potentially be missing?

Your discussion will greatly influence how we build this feature out.

**Comment by jaydrogers** [comment]
Original comment from @mojowill:

> I think there has to be an expectation from the user (myself being one) that there is going to be some manual work involved. 
> 
> - Spin init should create the base files but not actually run. (This is the problem when using spin init now. It fails when composer can't install packages due to missing php extensions)
> - spin creates a new example env file perhaps with just the new options it is setting like db connection details, Redis etc. 
> - the user is then responsible for modifying the docker files and env files to make it work. This would include any additional php extensions or system packages 
> 
> I'm writing this on my phone so probably will have things I can add once I'm actually at my desk trying to get this to work. 
> 
> I guess for me I'll init a new project then manually copy files across to my existing project and go from there.

**Reply by jaydrogers** [reply]
Another comment from @mojowill:

> I guess what you're doing currently is more a kin to sail:install in Laravel but with potentially the option to specify PHP versions and extensions?

**Reply by jaydrogers** [reply]
Hey @mojowill,

I haven't used `sail:install` before, I've been using Spin internally since 2018 😆

I just added this to the notes up top:

> ### Update the on-boarding screen
> 1. Choose PHP version
> 2. Ask if they have custom extensions
> 3. Detect if they have existing Dockerfiles
> 4. Give them documentation if they need to take any manual steps

Let me know your thoughts and if you have any other ideas!

**Comment by tonypiper** [comment]
How about breaking this down into more distinct migration scenarios, such as 'spin init' for sites using Laravel Herd?

Herd Pro made it very easy to get up and running on Mac. So I then went down the Forge and Envoyer route. But I still don't have CI/CD builds set up. I'd love to move to Docker but everytime I tinker I quickly get frustrated. Given that Herd has known features, versions, layout, perhaps it would be relatively easy migrate a Herd-based site to Spin.

**Reply by mojowill** [reply]
I think this is going to be the most common type of user (I'm one of them as well). 

We use Herd and Forge how do we get up and running with Spin and Spin Pro.

**Reply by tonypiper** [reply]
Also to say that Herd Pro costs $99. So this demographic are used to paying money for tooling that just works!

**Comment by luchtech** [comment]
Hi. I have an existing **Laravel Spin** project. Since I just bought the **Spin Pro**, I am sure that there are some changes that I need to do first. So how does converting Spin to Spin Pro looks like?

**UPDATE**: I tried running `spin init laravel-pro` on my existing Spin project, but I got this error since FilamentPHP requires the `ext-intl` PHP extension.

<img width="1340" alt="image" src="https://github.com/user-attachments/assets/9c202c3e-94e5-42fe-a15d-a4c1c5e983e7">

Maybe we can add a step to select `PHP extensions` to enable/install in the Spin Init process?

**Reply by thinkingcap** [reply]
Theres a section in the docs that explain how to accomplish that - 
https://getspin.pro/docs/guide/customizing-the-docker-image#installing-php-extensions

**Reply by mojowill** [reply]
> Theres a section in the docs that explain how to accomplish that - https://getspin.pro/docs/guide/customizing-the-docker-image#installing-php-extensions

Only when preparing your own Dockerfile. Spin init doesn't allow you to do this.

**Comment by jaydrogers** [comment]
I’ve been working on this all day today. I should hopefully have it ready by next week sometime 😅🤞

Here’s where I am at:

### Check and place files
- [x] Check for existing files
- [x] Prompt a warning if `--force` is not set
- [x] Display potential data loss may happen. Be sure to have backups of local databases and services. Be on a branch separate from your production code
- [x] Show warning we will automatically update your ENV files too

## Remove files
- [x] Remove `vendor` and `node_modules`
- [x] Remove `yarn.lock` and `composer.lock`
- [x] Any `Dockerfile*` in the project root
- [x] Remove `./.infrastructure`
- [x] Remove `docker-compose*`

### Select PHP Verison
- [x] Display a select box for PHP versions

### Prompt for PHP extensions
- [x] Show a link to default extensions
- [x] Show a link of available extensions
- [x] Enter a comma separated list

### Copy files
- [x] Copy over the base template
- [x] Update the blocks with the PHP version they selected
- [x] Add the PHP extensions
- [x] Process any extra "blocks"

### Update env
- [x] Update `.env` and `.env.example`

### Prepare project
- [x] Skip this if `--skip-dependency-install` is set
- [x] Install `serversideup/spin` package
- [x] Run `composer install`
- [x] Run `yarn install`
- [x] Run database migrations

**Reply by tonypiper** [reply]
also remove package-lock.json if it exists? (to reduce the WTFs potentially caused by not realising spin uses yarn instead of NPM)

**Reply by mojowill** [reply]
Yarn should be a choice but I should be able to continue using npm if I want to.

**Reply by jaydrogers** [reply]
Thanks for the feedback. All of this is done in my local version. I will push this up to Spin Pro soon.

Some components will be made available in the free version as well. 😃👍

**Reply by tonypiper** [reply]
As soon as pro can init an existing project I'll buy a licence! Hoping the intro pricing hasn't expired 😅

**Reply by jaydrogers** [reply]
@tonypiper: `spin init laravel-pro` is now available 😃 https://getspin.pro/docs/getting-started/add-to-an-existing-project

Intro pricing is still there 😉

**Comment by jaydrogers** [comment]
# Update
This is now available for testing on Spin Pro 🥳

https://github.com/serversideup/spin-template-laravel-pro/pull/24

![image](https://github.com/user-attachments/assets/d8c76bef-560d-4973-ac2e-b91a0194e649)

### For the open source version of `spin init laravel`
I will be adding a basic version of this soon 😃👍

**Comment by jaydrogers** [comment]
# 🚀 It's now available
You can now run `spin init laravel` or `spin init laravel-pro` to add Spin to your existing projects.

More content on this coming soon. Give it a spin and let me know if you run into any issues! It should be pretty straightforward to use from the onboarding wizard.
